doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -743,14 +749,125 @@ any expression, such as ``&0`` or ``&f(x)``.
+:func:`Base.cconvert(Csize_t, n) <cconvert>`.
+
+
+Here is a second example wrapping the corresponding destructor::
+
+    # The corresponding C signature is
+    #     void gsl_permutation_free (gsl_permutation * p);
+    function permutation_free(p::Ref{gsl_permutation})
+        ccall(
+            (:gsl_permutation_free, :libgsl), #name of C function and library
+            Void,                             #output type
+            (Ref{gsl_permutation},),          #tuple of input types
+            p                                 #name of Julia variable to pass in
+        )
+    end
+
+Here, the input ``p`` is declared to be of type ``Ref{gsl_permutation}``,
+meaning that the memory that ``p`` points to may be managed by Julia or by C.
+A pointer to memory allocated by C should be of type ``Ptr{gsl_permutation}``,
+but it is convertable using :func:`cconvert` and therefore can be used in the
+same (covariant) context of the input argument to a :func:`ccall`. A pointer to
+memory allocated by Julia must be of type ``Ref{gsl_permutation}``, to ensure
+that the memory address pointed to is valid and that Julia's garbage collector
+manages the chunk of memory pointed to correctly. Therefore, the
+``Ref{gsl_permutation}`` declaration allows pointers managed by C or Julia to
+be used.
+
+If the C wrapper never expects the user to pass pointers to memory managed by
+Julia, then using ``p::Ptr{gsl_permutation}`` for the method signature of the
+wrapper and similarly in the :func:`ccall` is also acceptable.
+
+
+Here is a third example passing Julia arrays::
+
+    # The corresponding C signature is
+    #    int gsl_sf_bessel_Jn_array (int nmin, int nmax, double x,
+    #                                double result_array[])
+    function sf_bessel_Jn_array(nmin::Integer, nmax::Integer, x::Real)
+        if nmax<nmin throw(DomainError()) end
+        result_array = Array(Cdouble, nmax-nmin+1)
+        errorcode = ccall(
+            (:gsl_sf_bessel_Jn_array, :libgsl), #name of C function and library
+            Cint,                               #output type
+            (Cint, Cint, Cdouble, Ref{Cdouble}),#tuple of input types
+            nmin, nmax, x, result_array         #names of Julia variables to pass in
+        )
+        if errorcode!= 0 error("GSL error code $errorcode") end
+        return result_array
+    end
+
+The C function wrapped returns an integer error code; the results of the actual
+evaluation of the Bessel J function populate the Julia array ``result_array``.
+This variable can only be used with corresponding input type declaration
+``Ref{Cdouble}``, since its memory is allocated and managed by
+Julia, not C. The implicit call to :func:`Base.cconvert(Ref{Cdouble},
+result_array) <cconvert>` unpacks the Julia pointer to a Julia array data
+structure into a form understandable by C.
+
+Note that for this code to work correctly, ``result_array`` must be declared to
```

Comment:
this isn't quite accurate – the gc-rooting is performed on the arguments without regards to the signature